### ISSUE TITLE
Get pull request number from action

### ISSUE THREAD

**Issue by estensen** [actions/checkout#58]
I'm trying to create an action that is triggered on a PR and tags a specified reviewer. For that I'm creating an action which uses the GitHub API to create a review request. To create this request I need the pull request number. The action is triggered by a PR, but I can't find the pull request number in the action logs.

**Comment by peter-evans** [comment]
You can extract the PR number from the environment variable `GITHUB_REF`. For `on: pull_request` workflows `GITHUB_REF` takes the format `refs/pull/:prNumber/merge`.

ref: https://help.github.com/en/github/automating-your-workflow-with-github-actions/events-that-trigger-workflows#pull-request-event-pull_request

**Comment by estensen** [comment]
Thanks! I also found this:
`pull_number=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")`

**Comment by n1ru4l** [comment]
**Javascript version**

```js
"use strict";

const parsePullRequestId = githubRef => {
  const result = /refs\/pull\/(\d+)\/merge/g.exec(githubRef);
  if (!result) throw new Error("Reference not found.");
  const [, pullRequestId] = result;
  return pullRequestId;
};

const main = async () => {
  const pullRequestId = parsePullRequestId(process.env.GITHUB_REF);
};

main().catch(err => {
  console.error(err);
  process.exitCode = 1;
});

```

**Comment by littlebtc** [comment]
Today I found that the PR number will be missing if you re-run the job for `on: pull_request` workflow. The only reliable way to get it is through $GITHUB_EVENT_PATH.

```javascript
const fs = require('fs')
const ev = JSON.parse(
  fs.readFileSync(process.env.GITHUB_EVENT_PATH, 'utf8')
)
const prNum = ev.pull_request.number
```

I think it is kind of a bug of GitHub Actions. 😭

**Comment by jpfeuffer** [comment]
I also think this is an important feature and it is a shame that this issue is closed "since you can hack your way around it". :(

**Comment by salvamomo** [comment]
In case this is still useful to anybody, a simple shell solution:

`PR_NUMBER=$(echo $GITHUB_REF | awk 'BEGIN { FS = "/" } ; { print $3 }')`

**Comment by leeyspaul** [comment]
Did anyone attempt to get the $GITHUB_REF but have a return with no PR number in the output? For example, my output is something like refs/head/master but has no pull number. Would love to see if anyone had a solution for this.

**Comment by leeyspaul** [comment]
> Did anyone attempt to get the $GITHUB_REF but have a return with no PR number in the output? For example, my output is something like refs/head/master but has no pull number. Would love to see if anyone had a solution for this.

Replying to myself. This seemed to work. ${{ github.event.pull_request.head.ref }} -shrug-!

**Comment by hunkim** [comment]
Is this problem fixed?

**Comment by astamato** [comment]
As I understood from the above, because there are several workarounds it's closed with no official method.

**Comment by ivasilko** [comment]
${{ github.event.pull_request.head.sha }}

**Comment by Bowenislandsong** [comment]
PR=$(IFS='/' read -r -a REF <<< "$GITHUB_REF" && echo ${REF[2]})

**Comment by kirosc** [comment]
> Did anyone attempt to get the $GITHUB_REF but have a return with no PR number in the output? For example, my output is something like refs/head/master but has no pull number. Would love to see if anyone had a solution for this.

Because it's not the first pull request. Instead, it's a `push` event.
GitHub should add the PR number to the environment variable.

**Comment by leeyspaul** [comment]
@kirosc I wrote a small post on retrieving the pull number based on an `issue` event.

If you use `${{github.event.number}}` you will get the PR number on pull request event.
If you use `${{github.event.issue.number}}` you will get the PR number on any issue event e.g. updates.

I hope that helps

**Comment by lukepighetti** [comment]
> @kirosc I wrote a small post on retrieving the pull number based on an `issue` event.
> 
> If you use `${{github.event.number}}` you will get the PR number on pull request event.
> If you use `${{github.event.issue.number}}` you will get the PR number on any issue event e.g. updates.
> 
> Not sure if that helps

Where is the documentation for these nested environment variables?

**Comment by polarathene** [comment]
It's not documented that well in the Github Actions/Workflow docs, but you'll find the `github.event.number` cited quite a bit around the web. I'm not sure where that's documented on official docs, unless `github.event` is vague and depends on context what properties it has.

They do have [official docs on pull request event](https://docs.github.com/en/developers/webhooks-and-events/github-event-types#pullrequestevent) which `github.event.pull_request.number` provides.

Issue event is similar, so `github.event.number` is a more generic approach (since on Github issues and PRs share the same incrementing number ID source for new issues/PRs).

**Comment by jeacott1** [comment]
@polarathene thank you! best answer here and the only one that also works on pull_request close!

**Comment by sangwoo-98** [comment]
Took me a serious amount of time to get here and see `github.event.number`,,,
Official documentation should be made ASAP.

**Comment by phil-lgr** [comment]
Sorry to add noise to this old issue but for those using the js toolkit to build custom js/ts actions, you can do:

```js
import * as github from '@actions/github';

const number = github.context.issue.number;
```

Source: https://github.com/actions/toolkit/blob/a1b068ec31a042ff1e10a522d8fdf0b8869d53ca/packages/github/src/context.ts#L55

**Comment by f0o** [comment]
So none of these actually work for me...

after `actions/checkout` I got:
```yaml
      - run: |
          echo github.event.number = ${{ github.event.number  }}
          echo github.event.pull_request.number = ${{ github.event.pull_request.number }}
          echo github.event.issue.number = ${{ github.event.issue.number }}
```

All of them are empty.

`$GITHUB_REF` returns the branch-name and not the PR Number.

So what's a reliable way of getting the PR-Number here? It can't be that hard seeing how PR/Merges are a core essence of git it's astonishing that this isn't available via a simple env-var by default..

**Comment by marians** [comment]
If there are still searchers passing by, https://github.com/marketplace/actions/find-current-pull-request solved this problem for me.

**Comment by dgroup** [comment]
One more way from stackoverflow: https://stackoverflow.com/a/69936361/6916890
```
 - name: Get Pull Request Number
        id: pr
        run: echo "::set-output name=pull_request_number::$(gh pr view --json number -q .number || echo "")"
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
and usage `${{ steps.pr.outputs.pull_request_number }}`.

**Comment by ar-mohammed** [comment]
> 
i have run the action/checkout but still i am not getting PR no

**Comment by polarathene** [comment]
@ar-mohammed it is available via the `github.event` context object, provided your workflow has been triggered by a `pull_request` event.

---

If the workflow is triggered by a different event, then you can use the `gh` CLI tool to make a query as was shown above. When you don't know what PR number/url/branch to reference, [the default is to return PR data for the current checked out branch](https://cli.github.com/manual/gh_pr_view).

That is what the above example is doing, followed by requesting the JSON field `number`  and `jq` (`-q`) expression to return that number, otherwise an empty string.

**Comment by ygots** [comment]
> If there are still searchers passing by, https://github.com/marketplace/actions/find-current-pull-request solved this problem for me.

It looks like it doesn't support comment-triggering events - I found this https://github.com/jwalton/gh-find-current-pr/issues/23

**Comment by rdhar** [comment]
Hi @estensen, 

To expand on @polarathene's answer about the API endpoint, here's a working snippet to get the issue number in both `push` and `pull_request` events within a GitHub Actions workflow by leveraging [actions/github-script](https://github.com/actions/github-script):

```js
steps:
  - uses: actions/github-script@v6
    id: get_issue_number
    with:
      script: |
        if (context.issue.number) {
          // Return issue number if present
          return context.issue.number;
        } else {
          // Otherwise return issue number from commit
          return (
            await github.rest.repos.listPullRequestsAssociatedWithCommit({
              commit_sha: context.sha,
              owner: context.repo.owner,
              repo: context.repo.repo,
            })
          ).data[0].number;
        }
      result-encoding: string
  - name: Issue number
    run: echo '${{steps.get_issue_number.outputs.result}}'
```

The script queries the [list labels for an issue](https://docs.github.com/en/rest/issues/labels#list-labels-for-an-issue) REST API endpoint via [octokit/rest.js](https://octokit.github.io/rest.js/v18#repos-list-pull-requests-associated-with-commit) client.

**Comment by mgaitan** [comment]
I've created an action that works for both `pull_request` and `merge_group` events. 
Check [mgaitan/gha-get-pr-number](https://github.com/mgaitan/gha-get-pr-number)

**Comment by sumanth-lingappa** [comment]
The below worked for me:
https://github.com/peter-evans/create-or-update-comment/issues/104#issuecomment-1140745353

**Comment by chriscz** [comment]
> `PR_NUMBER=$(echo $GITHUB_REF | awk 'BEGIN { FS = "/" } ; { print $3 }')`

This is problematic because the `refs` you see in `$GITHUB_REF` tend to have 3 or more path components:

- For a push to the main branch: `refs/heads/main`
- For a push to the develop branch: `refs/heads/develop`
- For a tag: `refs/tags/<tag-name>`
- For a pull request: `refs/pull/<pr-number>/merge`

I want this pattern to only match the last case. So I added a pattern to the last rule:

```
PR_NUMBER=$(echo $GITHUB_REF | awk 'BEGIN { FS = "/" } ; /^refs\/pull/ { print $3 }')
```
